### DISCUSSION TITLE
How does Muzei position the image (why is it offset to the left?!?)

### DISCUSSION THREAD

**Discussion by ianhanniballake** [romannurik/muzei#711] [Q&A] [answered]
Muzei doesn't have any controls over how to position the wallpaper. One some launchers, it appears like it is constantly offset to the left, even when you only have one screen. Yet, other third party launchers like [Nova Launcher](https://play.google.com/store/apps/details?id=com.teslacoilsw.launcher) and [Action Launcher](https://play.google.com/store/apps/details?id=com.actionlauncher.playstore) have a 'Disable wallpaper scrolling' option that seems to magically fix it.

When using a static image as my wallpaper, I am able to manually adjust the image's location, but I can't do that with Muzei.

How exactly does Muzei position the image? Why does changing the launcher have any effect?

**Comment by ianhanniballake** [comment, marked as answer]
Wallpapers and the APIs behind them are...odd, at best.

There's basically two entirely independent paths that have very little in common:
- using a static image as your wallpaper
- using an app that provides a live wallpaper

What you're describing is the static image case. Here, once an image is set, there's no app involved at all - the behavior is controlled by the Android OS (and any modifications your manufacturer has made). That means that some devices will have wallpaper scrolling even with a static image (that is the case for Pixel devices for instance) while perhaps that isn't the case for your device.

Muzei follows an entirely independent path using the 'live wallpaper' API, which involves implementing a [`WallpaperService.Engine`](https://developer.android.com/reference/android/service/wallpaper/WallpaperService.Engine). What happens is that the Android OS gives Muzei a [`SurfaceHolder`](https://developer.android.com/reference/android/view/SurfaceHolder) - a black box to draw the wallpaper into. Notably, the live wallpaper app has no control over the size of the Surface it is given nor any indication of what is currently visible to the user or not or how that Surface relates to the physical display.

So who does control the size of the Surface and how it relates to what is actually shown on the screen? That would be your launcher app. To set the size, they call [`suggestDesiredDimensions()`](https://developer.android.com/reference/android/app/WallpaperManager#suggestDesiredDimensions(int,%20int)) with *any* size. The most common is certainly one of the same height as the screen but wider, but other launchers change it in other ways to have a larger height as well (to support vertical pages in addition to horizontal pages...custom launchers get crazy).

So Muzei does the one thing that is consistent across every device and launcher - it centers the image in whatever space it is given, making sure the image extends from edge to edge in each direction.

It is then the launcher that is calling [`setWallpaperOffsets()`](https://developer.android.com/reference/android/app/WallpaperManager#setWallpaperOffsets(android.os.IBinder,%20float,%20float)) to effectively move its viewport around in that larger space. That's how wallpaper scrolling works - the launcher is calling a system API to change the offset in either direction.

That's why launchers like Nova Launcher and Action Launcher can offer an option to disable wallpaper scrolling - they're the ones that are setting the size and are the ones setting the offset as you swipe between screens. By setting the size to the same size as the screen and never calling any offsets, the image is always centered in the screen without scrolling.

Muzei, as a live wallpaper app, doesn't have the ability to override or change that part of how the Android APIs work. You're certainly not the first person to ask for it - in fact, this was issue #13 in Muzei, back in 2014. If anything had changed since then (and we continue to look at this every time a new release of Android is out), we definitely would have done it.

**Comment by Crono141** [comment]
Maybe this is a dumb question, but is there a way to update the wallpaper as a static image instead of using the live wallpaper API?  So the app could, for example, overwrite a static image that the Android OS is using as a static background.

**Reply by ianhanniballake** [reply]
Muzei is required to run as a live wallpaper, so no, this isn't possible, nor does would running it as a static image would help either (those are also sized to fit in whatever space the launcher provides).

**Comment by ghost** [comment]
I am still on Android 11, just out of curiosity, are things better on Android 12?

**Comment by Crono141** [comment]
Not that I know of. Reddit Wallpaper Changer doesn't have the weird
positioning issue which Muzei has, so Clearly there is a software solution
which could be implemented.

**Comment by baraa272** [comment]
kustom live wallpaper doesnt have this problem when applying an static image as live wallpaper why ?

**Comment by BitTim** [comment]
Wallpaper Changer by j4velin also has options as to what resolution to use (Screen resolution or launcher resolution) and also has the option of how the image should be handled (fill, fit or stretch), why not Muzei?
